---
title: "PPT_signatures_DIOLs"
author: "Nobuhiko Hamazaki"
date: "9/10/2020"
output: html_document
---

```{r message=FALSE}
library(tidyverse)
library(RColorBrewer) 
library(pheatmap)
```

# AT vs GC ratio 
```{r}

sel_df<- read.table(file = "table_AT_GC_dotplot_DIOLs.tsv", header = T)

##### gathering TPM table
long_df <- sel_df %>% gather(key = stage, value = TPM, -Geneid, -A_count, -T_count, -G_count, -C_count)



### AT counting
AT_high <- long_df[long_df$A_count + long_df$T_count > median(long_df$A_count + long_df$T_count) + 10,]
AT_low <- long_df[long_df$A_count + long_df$T_count < median(long_df$A_count + long_df$T_count) - 10,]

mZ_AT_high <- AT_high %>% na.omit() %>% as_tibble() %>%
   group_by(stage, add = F) %>%
   summarise(mean_TPM = mean(TPM))

mZ_AT_low <- AT_low %>% na.omit() %>% as_tibble() %>%
   group_by(stage, add = F) %>%
   summarise(mean_TPM = mean(TPM))

# dividing high values by low value
AT_hi_lo <- data_frame(Promoter_nucleotide = "AT", stage = mZ_AT_high$stage , high_per_low = (mZ_AT_high$mean_TPM-mZ_AT_low$mean_TPM) )

# scaling
AT_hi_lo$high_per_low <- scale(AT_hi_lo$high_per_low)

### GC counting
GC_high <- long_df[long_df$G_count + long_df$C_count > median(long_df$G_count + long_df$C_count) + 10,]

GC_low <- long_df[long_df$G_count + long_df$C_count < median(long_df$G_count + long_df$C_count) - 10,]

mZ_GC_high <- GC_high %>% na.omit() %>% na.omit() %>% as_tibble() %>%
   group_by(stage, add = F) %>%
   summarise(mean_TPM = mean(TPM))


mZ_GC_low <- GC_low %>%na.omit() %>% na.omit() %>% as_tibble() %>%
   group_by(stage, add = F) %>%
   summarise(mean_TPM = mean(TPM))

# dividing high values by low value
GC_hi_lo <- data_frame(Promoter_nucleotide = "GC", stage = mZ_GC_high$stage , high_per_low = (mZ_GC_high$mean_TPM - mZ_GC_low$mean_TPM) )

# scaling
GC_hi_lo$high_per_low <- scale(GC_hi_lo$high_per_low)


# merging two tables
ALL_hi_low <- rbind(AT_hi_lo, GC_hi_lo)


# Reorder factor
my_Order = c("ESCs", "IVD.D11", "IVD.D13", "DIOL.D5", "DIOL.D15", "DIOL.D25", "DIOL.S.D5", "DIOL.S.D14")

colnames(sel_df)

ALL_hi_low$stage <- factor(ALL_hi_low$stage, levels = my_Order)

ALL_hi_low %>% ggplot(aes(x=stage, y = Promoter_nucleotide)) +
  geom_point(aes(colour = high_per_low), size = 4) +
  scale_colour_gradient2(low = "steelblue", mid = "white", high = "red", midpoint = 0) +
  theme_classic() + 
  theme(axis.text.x = element_text(angle = 90,vjust = 0.5, hjust = 1, size = 12), axis.title.x = element_blank(), axis.title.y = element_blank(), legend.text = element_text(size = 8), legend.title = element_text(size = 10), plot.margin= unit(c(7, 1, 1, 1), "lines")) +
  labs(colour = "high/low")


```



# maternal gene expression
```{r}
# loading gene expression table
sel_df<- read.table(file = "table_AT_GC_dotplot_DIOLs.tsv", header = T)

# loading maternal gene list
mat <- readxl::read_xlsx("maternal_factors_list.xlsx")

# joining 2 tables
mat_gene <- inner_join(x = mat, y = sel_df, by = c("Genes"="Geneid"))

mat_gene_sel <- mat_gene %>% dplyr::filter(Function == "Preimplantation_development" | Function == "DNA_methylation")


# selecting cols and filtering genes based on tpm
mat_expr <- mat_gene_sel %>% dplyr::select( 
   contains("ESCs"),
   starts_with("IVD.D11"),
   starts_with("IVD.D13"),
   contains("DIOL.D5"),
   contains("DIOL.D15"),
   contains("DIOL.D25"),
   contains("DIOL.S.D5"),
   contains("DIOL.S.D14")
   )  %>% as.data.frame()

# add row names
row.names(mat_expr) <- mat_gene_sel$Genes

# heatmap
mat_expr %>% 
  pheatmap(cluster_cols = F, cluster_rows = F, scale = "row", cellwidth = 10, cellheight = 10)



```





# X/A ratio
# Calcurating X chromosomal expression
```{r out.height= 1000, out.width= 8000}
# Merging with bed table for ataching chr information

# Laoding information from bed file
bed <- readr::read_delim("RefSeq_name.bed", delim = "\t",col_names = c("chr","Start","End","Geneid", "Strand")) %>% dplyr::filter(!grepl("Un|random",chr))


# loading gene expression table
sel_df<- read.table(file = "table_AT_GC_dotplot_DIOLs.tsv", header = T) %>% dplyr::select(-A_count, -T_count, -G_count, -C_count)



# joining by inner_join and removing dups by distinct
tpm.bed <- dplyr::inner_join(x= sel_df, y=bed, by = "Geneid") %>%
  dplyr::distinct(Geneid, .keep_all = T)


# gathering for long format
mgxp <- tpm.bed  %>% 
  gather(stage, TPM, -Geneid, -chr, -Start, -End, -Strand)


 
# X/A ratio calculation
X.link <- mgxp[mgxp$chr=="chrX",]

A.link <- mgxp %>% 
  dplyr::filter(chr != "chrX" & chr!= "chrY")

# Calc. sum values
X_sum <- X.link[,6:7] %>% group_by(stage) %>% summarise_all(funs(sum))
A_sum <- A.link[,6:7] %>% group_by(stage) %>% summarise_all(funs(sum))

# Calc. X/A ratio
XperA_sum <- X_sum[,2]/A_sum[,2]
# adding stage column
XAsum <- data.frame(XperA_sum, A_sum[,1])

XAsum$TPM <-  XAsum$TPM * 19

# Reorder factor
my_Order = c("ESCs", "IVD.D11", "IVD.D13", "DIOL.D5", "DIOL.D15", "DIOL.D25", "DIOL.S.D5", "DIOL.S.D14")

 # reorder
XAsum$stage <- factor(XAsum$stage, levels = my_Order)
factor(XAsum$stage)

my_cols <- c(rep("black", 5), "red",  "blue", "black")


ggplot(XAsum, aes(x=stage, y=TPM)) +
  geom_point(size = 4) +
  theme_classic(base_size = 12, base_family = "Helvetica") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ylim(c(0, 1.2)) +
  ggtitle("X/A ratio") +
  ylab("X/A ratio") +
  geom_abline(col = "red", intercept = 0.5, slope = 0, size = 0.1) +
  xlab("")
         

```







# LTR expression
```{r}
# loading repeat expression tables
df <- read.table("table_repeat_DIOLs.tsv", header = T, check.names = F)

# gather
g_rep_rpm <- df %>%  gather(key = stage, value = RPM, -Geneid, -Repeat_type)

# removing charactors after "rep"
g_rep_rpm$stage <- g_rep_rpm$stage %>% str_sub(start = 1, end = -6) 

# average table
mean_rpm <- g_rep_rpm %>% as_tibble() %>%
  group_by(Geneid, Repeat_type, stage) %>%
  summarise(mean_RPM =  mean(RPM)) %>%
  tidyr::spread(key = stage, value = mean_RPM)


# repeat select
sel_mean_rpm1 <- mean_rpm %>% dplyr::filter(Geneid == "LTR", Repeat_type == "Class") 
sel_mean_rpm2 <- mean_rpm %>% dplyr::filter(Geneid == "ERVK" | Geneid == "ERVL-MaLR")
sel_mean_rpm3 <- mean_rpm %>% dplyr::filter(Geneid == "MT-int" | Geneid == "MTA_Mm-int" | Geneid == "RLTR10-int")
sel_mean_rpm <- bind_rows(sel_mean_rpm1, sel_mean_rpm2,sel_mean_rpm3)

# adding row name
mean_mx_rpm <- as.matrix(sel_mean_rpm[,c(-1,-2)])
rownames(mean_mx_rpm) <- sel_mean_rpm$Geneid

# reorder
mean_mx_rpm2 <- mean_mx_rpm[,c(6,7,8,3,1,2,5,4)]


# heatmap
mean_mx_rpm2 %>%  
  pheatmap( cluster_cols = F, cluster_rows = F, cellwidth = 10, cellheight = 10, scale = "row",  gaps_row = c(1,3))




```






